cpp/src/arrow/memory_pool.h

```diff
@@ -56,6 +58,33 @@ class ARROW_EXPORT MemoryPool {
   /// The number of bytes that were allocated and not yet free'd through
   /// this allocator.
   virtual int64_t bytes_allocated() const = 0;
+
+  /// Return peak memory allocation in this memory pool
+  ///
+  /// \return Maximum bytes allocated. If not known (or not implemented),
+  /// returns -1
+  int64_t max_memory() const;
+
+ protected:
+  MemoryPool();
+  std::atomic<int64_t> max_memory_;
```

Comment:
Unused yet and probably not simple to update the maximum in a single atomic operation.